plugins/regex_revalidate/regex_revalidate.cc

```diff
@@ -0,0 +1,754 @@
+
+        if (vecinv) {
+          time_t const date    = get_date_from_cached_hdr(txn);
+          time_t const timenow = time(nullptr);
+          char *url            = nullptr;
+          int url_len          = 0;
+
+          for (Invalidate const &inv : *vecinv) {
+            if (date <= inv.epoch && timenow <= inv.expiry) {
+              if (nullptr == url) {
+                url = TSHttpTxnEffectiveUrlStringGet(txn, &url_len);
+              }
+
+              if (nullptr == url || 0 == url_len) {
+                break;
+              } else if (inv.matches(url, url_len)) {
+                TSHttpTxnCacheLookupStatusSet(txn, inv.new_result);
+                DEBUG_LOG("Forced revalidate - %.*s", url_len, url);
+                break;
+              }
+            }
+          }
+
+          if (nullptr != url) {
+            TSfree(url);
+          }
+        }
+      }
+    }
+    break;
+  default:
+    break;
+  }
+
+  TSHttpTxnReenable(txn, TS_EVENT_HTTP_CONTINUE);
+  return 0;
+}
+
+} // namespace
+
+namespace
+{
+void
+setup_config_cont(PluginState *const pstate)
+{
+  if (nullptr == config_cont) {
+    DEBUG_LOG("creating config continuation");
+
+    // create reusable configuration file regex
+    constexpr char const *const expr = "^([^#].+?)\\s+(\\d+)(\\s+(\\w+))?\\s*$";
+    bool const cstat                 = Invalidate::config_re.compile(expr);
+
+    if (!cstat) {
+      ERROR_LOG("setup_config_cont: Unable to compile config_re, disabling plugin");
+      return;
+    }
+
+    // set up the config continuation
+    config_cont = TSContCreate(config_handler, TSMutexCreate());
+    TSContDataSet(config_cont, static_cast<void *>(pstate));
+
+    // constantly run this for rule grooming
+    TSContScheduleOnPool(config_cont, CONFIG_TMOUT, TS_THREAD_POOL_TASK);
+
+    // also register this continuation with any config reloads
+    TSMgmtUpdateRegister(config_cont, PLUGIN_NAME);
+
+    // and also register as a lifecycle hook
+    TSLifecycleHookAdd(TS_LIFECYCLE_MSG_HOOK, config_cont);
+  }
+}
+
+} // namespace
+
+void
+TSPluginInit(int argc, char const *argv[])
+{
+  DEBUG_LOG("Starting plugin init");
+
+  TSPluginRegistrationInfo info;
+  info.plugin_name   = PLUGIN_NAME;
+  info.vendor_name   = "Apache Software Foundation";
+  info.support_email = "dev@trafficserver.apache.org";
+
+  if (TSPluginRegister(&info) != TS_SUCCESS) {
+    ERROR_LOG("Global plugin registration failed");
+    return;
+  } else {
+    DEBUG_LOG("Global plugin registration succeeded");
+  }
+
+  if (TS_VERSION_MAJOR != TSTrafficServerVersionGetMajor()) {
+    ERROR_LOG("Plugin requires Traffic Server %d", TS_VERSION_MAJOR);
+    return;
+  }
+
+  PluginState *const pstate = new PluginState;
+  if (!pstate->fromArgs(argc, argv)) {
+    ERROR_LOG("Remap plugin registration failed");
+    delete pstate;
+    return;
+  }
+
+  setup_config_cont(pstate);
+
+  TSCont const main_cont = TSContCreate(main_handler, nullptr);
```

Comment:
Don't the two continuations need the same mutex since they both access pstate?

Reply:
i believe that wouldn't allow the invalidation and the config continuations to run asynchronously.  That's what the atomic locks  for pstate does.  I'm guessing this might break if global plugin reloading is enabled?